devnotes/sbe.md

```diff
@@ -0,0 +1,127 @@
+new version of SBE and HBBL is able to start Hostboot it won't get updated.
+
+## Firmware
+
+As mentioned, this chip does not use the same ISA as PPC405. Because of that it
+is compiled by customised [gcc](https://github.com/open-power/ppe42-gcc) and
+[binutils](https://github.com/open-power/ppe42-binutils) forks.
+
+Firmware is located in SEEPROM, along with HBBL, Secure Boot keys and some
+metadata like version numbers etc. ECC is used for whole SEEPROM in a way that
+no ECC block crosses 64kB, instead a padding (64kB % 9 = 7 bytes) is added at
+the end of each 64kB part. It means that the actual size of code and data in
+SEEPROM is (64kB - 7) * 4 * 8 / 9 = 232992 bytes = ~227,5 kilobytes.
+
+All offsets in the image should be applied to the image with ECC and padding
+removed. This can be done with the following commands, assuming 64kB blocks are
+saved into `_seeprom0` through `_seeprom3` files:
+
+```
+head -c-7 -q _seeprom? > seeprom.bin.ecc
+ecc -R seeprom.bin.ecc -p -o seeprom.bin
+```
+
+To add ECC bytes and split it back:
+
+```
+ecc -I seeprom.bin -p -o seeprom.bin.ecc
+split -d -a 1 -b 65529 seeprom.bin.ecc --filter='dd bs=65536 conv=sync of=$FILE' _seeprom
+```
+
+SEEPROM begins with [P9XipHeader structure](https://github.com/open-power/sbe/blob/master/src/import/chips/p9/xip/p9_xip_image.h#L368).
+A part of that structure is an array of [P9XipSection structures](https://github.com/open-power/sbe/blob/master/src/import/chips/p9/xip/p9_xip_image.h#L280),
+one of such structures is the one describing HBBL. It is located at offset 0xE8
+relative to the beginning of SEEPROM (0x105 with ECC). For the current
+`07-25-2019` branch of SBE HBBL is 20kB big, located at offset 0x2F008. With ECC
+and padding it lands entirely in last quarter of SEEPROM. Note that Secure ROM
+(SHA512 algorithm) is a big part of that 20kB.
+
+## Reading SEEPROM from BMC
+
+There are at least two ways of reading (and possibly writing) SEEPROM from BMC.
+We can either use a kernel driver to read directly by I2C bus, or do this
+manually by writing and reading appropriate SCOM registers.
+
+#### I2C
+
+SEEPROM is split into four 64kB parts, each can be accessed under a separate I2C
+address: 0x54, 0x55, 0x56, 0x57. SEEPROM for CPU0 is located under bus 0, CPU1
+under bus 1.
+
+All parts are mounted by default, unfortunately wrong kernel driver is used and
+only the first 32kB of each block is accessible. The easiest way of using a
+proper driver is to create a new device. Because we cannot easily remove the
+device that was created automatically we have to use 10 bit addressing and set a
+bit that is ignored by SEEPROM, otherwise kernel won't let us use the same
+address as used by existing driver. To add a new device:
+
+```
+echo 24c512 0xa0d4 > /sys/bus/i2c/devices/i2c-0/new_device
+```
+
+`0xa0..` tells to use 10 bits addressing mode, `0x..d4` is the address. Note
+that its 7 lowest bits are the same as in `0x54`. This must be repeated for each
```

Comment:
Does it have to be repeated for addresses 0x55, 0x56 and 0x57? If yes, add the same commands for them:
```
echo 24c512 0xa0d4 > /sys/bus/i2c/devices/i2c-0/new_device
echo 24c512 0xa0dX > /sys/bus/i2c/devices/i2c-0/new_device
...
```

Reply:
Only if one wants to read/write all of SEEPROM at once. I wanted to only describe why these exact addresses are used here, full binding is done in script mentioned in the next paragraph.

I wanted users to have to think what they're doing instead of just copying commands. Playing with SEEPROM may be dangerous. This is also the reason why the writing half of the script is commented out.